### PR TITLE
feat: config-driven pipeline with spatial batching and multi-dataset support

### LINKED ISSUE #9: Config-driven pipeline with spatial batching and multi-dataset support
## Summary

Implement the config-driven 5-stage pipeline from design.md section 4, replacing the spike/one-off script approach with a general system driven by YAML configuration files.

## Motivation

The current spike (`#6`) demonstrated feasibility but hard-coded dataset access and processing logic. Users configure *what* to compute via YAML; the pipeline determines *how*. Adding a new dataset should require only YAML changes (registry entry + pipeline config), no code changes.

## Key Features

- **Rich config schema** (Pydantic): TargetFabricConfig, DomainConfig, DatasetRequest, OutputConfig, ProcessingConfig with batch_size
- **Dataset registry** (`configs/datasets.yml`): Maps dataset names to access strategies (stac_cog, converted_zarr, local_tiff) with categorical/continuous/derived variable support
- **Spatial batching** (KD-tree recursive bisection): Groups catchments into spatially contiguous batches for memory-efficient 10m resolution processing on desktop hardware
- **STAC COG data access**: Planetary Computer integration with multi-tile mosaicing
- **Terrain derivation**: Horn (1981) slope/aspect from numpy.gradient, no external raster deps
- **gdptools ZonalGen processing**: exactextract engine for continuous + categorical variables
- **5-stage pipeline orchestrator**: Fabric → Datasets → Batched Processing → SIR assembly → NetCDF/Parquet output
- **CF-1.8 compliant SIR** (xr.Dataset) with provenance attributes

### New modules
- `batching.py` — KD-tree recursive bisection (~50 lines)
- `dataset_registry.py` — YAML-driven registry with VariableSpec/DerivedVariableSpec
- `data_access.py` — STAC COG fetch, terrain derivation, GeoTIFF I/O
- `pipeline.py` — 5-stage orchestrator with batch loop + CLI entry point

### Modified modules
- `config.py` — Rich Pydantic schema replacing minimal spike
- `processing.py` — gdptools ZonalGen wrapper (continuous + categorical)
- `__init__.py` — Updated public API exports

### Config files
- `configs/datasets.yml` — Registry: 3DEP (STAC), POLARIS (Zarr), NLCD (TIFF)
- `configs/delaware_terrain.yml` — MVP pipeline config

### Tests
- 61 unit tests across 6 test files (all pass, 65% coverage)
- Known-answer terrain tests (flat surface, 45° tilted planes, cardinal aspects)
- Spatial batching tests (locality, balance, edge cases)
- Pipeline stage tests (fabric, dataset resolution, SIR assembly, CF metadata)

## Dependencies

- New `stac` optional group: pystac-client, planetary-computer, rioxarray  
- Added `exactextract` to `gdp` group
- New pixi features for `gdp` and `stac` with conda-forge overrides

### PR DESCRIPTION
**Pull request by rmcd-mscb**
## Summary

- Replace spike/one-off script approach with a general **config-driven 5-stage pipeline** from design.md section 4
- Add **spatial batching** (KD-tree recursive bisection) enabling 10m resolution processing on desktop hardware
- Add **dataset registry** supporting multiple access strategies (STAC COG, Zarr, local TIFF) and variable types (continuous, categorical, derived)
- Add **terrain derivation** (Horn slope/aspect), **STAC COG access** (Planetary Computer), and **gdptools ZonalGen** processing
- 61 tests across 6 files, all passing at 65% coverage

### New modules
| Module | Purpose |
|--------|---------|
| `dataset_registry.py` | YAML-driven registry with VariableSpec/DerivedVariableSpec |
| `batching.py` | KD-tree recursive bisection (~50 lines core) |
| `data_access.py` | STAC COG fetch, terrain derivation, GeoTIFF I/O |
| `pipeline.py` | 5-stage orchestrator with batch loop + CLI |

### Config files
| File | Purpose |
|------|---------|
| `configs/datasets.yml` | Registry: 3DEP (STAC), POLARIS (Zarr), NLCD (TIFF) |
| `configs/delaware_terrain.yml` | MVP pipeline config for Delaware basin |

### Modified modules
- `config.py` — Rich Pydantic schema (batch_size, engine, domain types)
- `processing.py` — gdptools ZonalGen wrapper (continuous + categorical)
- `__init__.py` — Updated public API exports
- `docs/design.md` — New section 11: MVP implementation decisions

## Test plan

- [x] `pixi run -e dev check` passes (lint, format, typecheck, 61 tests)
- [x] `pixi run -e dev pre-commit` passes (all 10 hooks)
- [ ] End-to-end: `python -m hydro_param.pipeline configs/delaware_terrain.yml` (requires network + STAC data)

Closes #9

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: docs/design.md [hunk 1]
```diff
@@ -1218,13 +1218,20 @@ Failed HRU log: ./failed_hrus.csv
...
+
+- [x] Config file schema design (declarative YAML structure) — §11.6
+- [x] Test domain selection and MVP scope (Delaware River Basin) — §11.1
+- [ ] Config schema implementation (Pydantic models + YAML loader)
+- [ ] Dataset registry implementation (YAML schema + variable resolution)
+- [ ] Spatial batching implementation (KD-tree recursive bisection) — §11.2
+- [ ] STAC COG data access (Planetary Computer integration) — §11.4
+- [ ] gdptools ZonalGen + exactextract integration — §11.5
+- [ ] Pipeline orchestrator (5-stage batch loop) — §11.7
+- [ ] SIR output as xr.Dataset with CF-1.8 metadata
```
[the remark this answers is not in the document]
Comment (rmcd-mscb) [reply docs/design.md]: Fixed in 4c729bd — marked all implemented items as checked.

File: src/hydro_param/processing.py [hunk 2]
```diff
@@ -1,48 +1,151 @@
...
+            len(result_df),
+            list(result_df.columns),
         )
 
+        # For continuous variables, select requested statistics
+        if not categorical:
+            available = set(result_df.columns)
+            selected = [s for s in statistics if s in available]
+            if selected:
+                result_df = result_df[selected]
```
[the remark this answers is not in the document]
Comment (rmcd-mscb) [reply src/hydro_param/processing.py]: Fixed in 4c729bd — now logs a warning when requested statistics are not in the result. Chose warning over raising because gdptools may return different stat column names depending on engine, and we don't want to break valid runs over naming.

File: src/hydro_param/dataset_registry.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    sign: str | None = None
+    # Zarr / local fields
+    source: str | None = None
+    # Common fields
+    crs: str = "EPSG:4326"
+    variables: list[VariableSpec] = []
+    derived_variables: list[DerivedVariableSpec] = []
+    category: str = ""
+    temporal: bool = False
+
```
[the remark this answers is not in the document]
Comment (rmcd-mscb) [reply src/hydro_param/dataset_registry.py]: Fixed in 4c729bd — added a `@model_validator` that enforces `catalog_url`+`collection` for stac_cog and `source` for local_tiff. Two new tests added.

File: src/hydro_param/pipeline.py [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+    logger.info("=" * 60)
+
+    # Stage 1: Resolve target fabric
+    fabric = stage1_resolve_fabric(config)
+
+    # Spatial batching
+    fabric = spatial_batch(fabric, batch_size=config.processing.batch_size)
+
+    # Stage 2: Resolve source datasets
+    resolved = stage2_resolve_datasets(config, registry)
```
[the remark this answers is not in the document]
Comment (rmcd-mscb) [reply src/hydro_param/pipeline.py]: Good catch. For the MVP, `resolve_bbox()` is scaffolding for future use — the fabric is currently pre-filtered by the download script. Added a TODO in `run_pipeline()` to apply domain-based subsetting in a follow-up. The bbox will be used to clip the fabric before batching once we support HUC/gage domain types.

File: src/hydro_param/pipeline.py [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+            da = derive_fn(source_da, method=var_spec.method)
+        else:
+            # Raw variable: fetch directly
+            if entry.strategy == "stac_cog":
+                da = fetch_stac_cog(entry, bbox)
+            else:
+                raise NotImplementedError(f"Strategy '{entry.strategy}' not yet supported")
+            # Cache for potential derived variable reuse
+            source_cache[var_spec.name] = da
+
```
[the remark this answers is not in the document]
Comment (rmcd-mscb) [reply src/hydro_param/pipeline.py]: Agreed — all current datasets are single-band, but this will matter for multi-band rasters. Added a TODO at the fetch site in `_process_batch()` to pass `var_spec.band` through to the fetch routine.
